src/fsharp/FSharp.Core/array.fs

```diff
@@ -1270,6 +1270,88 @@ namespace Microsoft.FSharp.Collections
                 let count' = Operators.min count len
                 Microsoft.FSharp.Primitives.Basics.Array.subUnchecked 0 count' array
 
+        [<CompiledName("RemoveAt")>]
+        let removeAt (index: int) (source: 'T[]) : 'T[] =
+            checkNonNull "source" source
+            if index < 0 || index >= source.Length then invalidArg "index" "index must be within bounds of the array"
+            
+            let length = source.Length - 1
+            let result = Microsoft.FSharp.Primitives.Basics.Array.zeroCreateUnchecked length
+            Array.Copy(source, result, index)
```

Comment:
Well, I like to quote Steven Toub here: it's harmful to add such checks because they actually regress perf in most used scenario, which is when array is not empty.

Reply:
Hmmm I don't know if it's the least used case.  I mean, it could be that this is often used on small lists where the insert is at front or back.  And Array.Copy will have a bunch of null checks, overlap checks etc. (e.g. copying same array to itself) I expect. 

Anyway this is what perf benchmarking is for, to check the difference :)  The benchmarks should I guess include both large array and small array tests (size 0, 1, 2, 3), including insert-at-start and insert-at-end

Reply:
i added the check for all array.copy calls